service/gcs/core/gcs/gcs.go

```diff
@@ -430,26 +423,36 @@ func (c *gcsCore) RunExternalProcess(params prot.ProcessParameters, stdioSet *co
 		cmd.SetStdout(console)
 		cmd.SetStderr(console)
 	} else {
-		if stdioOptions.CreateIn {
-			// Stdin uses cmd.StdinPipe() instead of cmd.Stdin because cmd.Wait()
-			// waits for cmd.Stdin to return EOF. We can only guarantee an EOF by
-			// closing the pipe after the process exits.
-			cmdStdin, err := cmd.StdinPipe()
+		// Get the underlying files for each stdio socket and use them directly
+		// for the stdio fds for the new process. Close these files as soon as
+		// the process is launched so that only the child process has open
+		// descriptors to these sockets.
+		if stdioSet.In != nil {
+			fileIn, err = stdioSet.In.File()
 			if err != nil {
-				return -1, errors.Wrap(err, "failed to get stdin pipe for command")
+				return -1, errors.Wrap(err, "failed to dup stdin socket for command")
 			}
-			wg.Add(1)
-			go func() {
-				io.Copy(cmdStdin, stdioSet.In)
-				cmdStdin.Close() // Notify the process that there is no more input.
-				wg.Done()
-			}()
+			defer fileIn.Close()
+			defer stdioSet.In.Close()
+			cmd.SetStdin(fileIn)
 		}
-		if stdioOptions.CreateOut {
-			cmd.SetStdout(stdioSet.Out)
+		if stdioSet.Out != nil {
+			fileOut, err = stdioSet.Out.File()
+			if err != nil {
+				return -1, errors.Wrap(err, "failed to dup stdout socket for command")
+			}
+			defer fileOut.Close()
+			defer stdioSet.Out.Close()
+			cmd.SetStdout(fileOut)
 		}
```

Comment:
Since it's an file, you save one fd. The old way would do a ```pipe(2)``` and then send data through the pipe. In this case, it just does a ```dup2```.